### DISCUSSION TITLE
Directory specific 404 page

### DISCUSSION THREAD

**Discussion by shreyas-jadhav** [vercel/next.js#23764] [Ideas]
### Describe the feature you'd like to request

A custom 404 page resides at ```pages/404.js``` , but if we are in directory ```pages/users``` and user accesses a dynamic page like for example pages/users/[id].js . So if ```getStaticProps()``` returns ```notFound```, then we should get a custom 404 page saying that user doesn't exists which resides at ```pages/users/404.js```, instead of Page not Found from ```pages/404.js```  . 

Similarly if we have another directory for example, ```pages/settings``` and user navigates to some unknown page like ```pages/settings/unknown``` for example, this should return a 404 page from pages/settings saying custom 404 page like _"No such setting page exists, try something else, etc."_

### Describe the solution you'd like

Allow sub-directories in pages folder to have **independent 404 pages**, instead of using same 404 page (```page/404.js```).

### Describe alternatives you've considered

1. While using ```getStaticProps()``` return  ```redirect: { destination: "./404" }``` 
(but this also changes the url to ```/users/404```, and cannot retain original URL.)

2. While using ```getStaticProps()``` Using custom 404 component, check is the data from props exists, render 404 component. (Not ideal if we have many pages)

**Comment by jpomykala** [comment]
Maybe you could just use default 404.js page and adjust the view based on the current path.

**Comment by shreyas-jadhav** [comment]
Yup, i can do it for now. But I think it's more cleaner to have directory specific pages.
Imagine implementing all the components in a single file and render each one based on path, that would be so complicated.
And what if we want to use SSG for the 404 page?
